Address Pool work with key/address fingerprints

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#901 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have created an interface for performing hard and soft key derivation on keys 
- [x] I have temporarily removed 'KnownAddresses' interface, to be replaced with a DB call
- [x] I have updated `AddressPool` and `SeqState` to no longer require a `ShelleyKey` nor `Address`, but instead, use an abstract `key` type and track `KeyFingerprint` (which can be obtained from both Shelley and Byron keys).
- [x] I have revised how keys get serialized / unserialize to the database to have cleaner interfaces

# Comments

<!-- Additional comments or screenshots to attach if any -->

:warning: base is `KtorZ/900/inspect-address-tests` :warning: 

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
review interface names for 'InspectAddress'

KtorZ committed:
define interface for hard and soft derivation for keys

KtorZ committed:
remove 'KnownAddresses' interfaces, to be replaced with DB calls at call-site

KtorZ committed:
draft: allow computing a key fingerprint from an address public key

KtorZ committed:
WIP Review 'AddressPool' to work with key fingerprints instead of addresses!

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -144,6 +151,34 @@ import qualified Data.Text.Encoding as T
+    toText = toTextFromBoundedEnum SnakeLowerCase
+
+instance FromText ChangeChain where
+    fromText = fromTextToBoundedEnum SnakeLowerCase
```
Moved from `AddressDerivation.ShelleyKey`

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -394,36 +374,32 @@ nextChangeIndex pool (PendingIxs ixs) =
-        <> indentF 4 (build extP)
-        <> indentF 4 ("Change indexes: " <> indentF 4 chgsF)
-      where
-        chgsF = blockListF' "-" build (pendingIxsToList chgs)
```
I've removed 'Buildable' instances as a lazy way to update them for now :grimacing: ... If they were really needed somewhere, I'll actually update them.. otherwise.. bye.

KtorZ committed:
Make 'KeyFingerprint' a top-level type and rename 'InspectAddress' to 'MkKeyFingerprint'

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs:
```diff
@@ -491,8 +575,22 @@ class WalletKey key => PersistKey (key :: Depth -> * -> *) where
         -> Either String (key 'RootK XPrv, Hash "encryption")
 
 -- | Access constituants of an address.
```
```suggestion
-- | Access constituents of an address.
```

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -221,84 +187,49 @@ unsafeGenerateKeyFromSeed (Passphrase seed, Passphrase gen) (Passphrase pwd) =
+            return $ ShelleyKey addrXPub
+      where
+        errWrongIndex = error $
+            "Cardano.Wallet.Primitive.AddressDerivation.deriveAddressPublicKey \
```
```suggestion
            "Cardano.Wallet.Primitive.AddressDerivation.Shelley.deriveAddressPublicKey \
```

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -440,21 +371,36 @@ decodeShelleyAddress bytes = do
-        | otherwise =
-            error "InspectAddress: tried to inspect an incompatible address"
+        | otherwise = error $ unwords
+            [ "InspectAddress.getPaymentKey was given an invalid 'ShelleyKey'"
```
```suggestion
            [ "InspectAddress.paymentKeyFingerprint was given an invalid 'ShelleyKey'"
```

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDerivation/Shelley.hs:
```diff
@@ -440,21 +371,36 @@ decodeShelleyAddress bytes = do
-        | otherwise =
-            error "InspectAddress: tried to inspect an incompatible address"
+        | otherwise = error $ unwords
+            [ "InspectAddress.getDelegationKey was given an invalid 'ShelleyKey'"
```
```suggestion
            [ "InspectAddress.delegationKeyFingerprint was given an invalid 'ShelleyKey'"
```

paweljakubas commented on lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Sequential.hs:
```diff
@@ -193,27 +175,25 @@ defaultAddressPoolGap =
 -- >>> mkAddressPool xpub gap changeChain mempty
 -- AddressPool { }
-data AddressPool (network :: NetworkDiscriminant) (chain :: ChangeChain) = AddressPool
+data AddressPool (chain :: ChangeChain) (key :: Depth -> * -> *) = AddressPool
```
:100:

paweljakubas reviewed (approved):
except some small edit suggestions looks good so far.

paweljakubas commented on lib/core/test/unit/Cardano/Wallet/Primitive/AddressDerivation/ShelleySpec.hs:
```diff
@@ -153,3 +200,21 @@ instance {-# OVERLAPS #-} Arbitrary (Passphrase "seed") where
+newtype InvalidAddress = InvalidAddress Address deriving (Eq, Show)
+
+instance Arbitrary InvalidAddress where
+    arbitrary = InvalidAddress . Address . BS.pack
```
maybe it would be good to cover more here. We want to check that 
1. `KeyFingerprint "payment" ShelleyKey` will accept either `addrSingleSize` or `addrGroupedSize` length addresses. Otherwise error.
2. `KeyFingerprint "delegation" ShelleyKey` will accept `addrGroupedSize` length address. When `addrSingleSize` length address comes it will return Nothing. Otherwise, error

Now we expose it with `addrSingleSize + addrGroupedSize` size address. So just one address size. Maybe we can be more diverse here ?

KtorZ committed:
define common interface for key serialization and make address pool key agnostic

KtorZ committed:
propagate changes from previous commit to tests and benchmarks

KtorZ committed:
Only store the already pre-calculated HD passphrase in the rnd state

This to avoid the ugly race condition in the database when trying to
read a random state for which there's not yet any private key attached.

KtorZ committed:
adjust comments in the AddressDerivation module

KtorZ committed:
cover a broader spectrum of invalid addresses in KeyFingerprint properties

KtorZ committed:
remove unused pragmas
